[BEAM-1779] Port UnboundedSourceWrapperTest to use Flink operator test harness

aljoscha opened the pull request.

aljoscha commented:
Run Flink RunnableOnService

davorbonaci commented:
hey @aljoscha, it this PR still valid/needed/etc? Can I help somehow perhaps?

aljoscha commented:
@davorbonaci It's not hight priority since it doesn't fix any immediate problem. I think it would still be good to have because it brings the testing in line with how it should be done and sets a good example. I probably have to rebase this once the update to Flink 1.3 is in: #3284

iemejia commented:
@aljoscha please remember to change the name of the PR from FLINK-1779 to BEAM-1779 as well as the commit.

aljoscha commented:
I actually found a bug while rebasing this on top of master: https://issues.apache.org/jira/browse/BEAM-2407

cc: @JingsongLi could you please also take a look?

aljoscha commented:
Turns out there is yet another bug that should be catched by these tests: https://issues.apache.org/jira/browse/BEAM-2408

I fixed it and added a test for watermark emission.

CC: @iemejia This is a pre-requisite for the "close on +Inf watermark" PR now.

aljoscha commented:
Run Flink ValidatesRunner

aljoscha commented:
@tgroh Could you maybe have a look at the added tests. And could you maybe review, @JingsongLi?

JingsongLi commented:
Let me take a look.

JingsongLi commented on runners/flink/src/main/java/org/apache/beam/runners/flink/translation/types/CoderTypeSerializer.java:
```diff
@@ -139,24 +140,28 @@ public TypeSerializerConfigSnapshot snapshotConfiguration() {
-    private Coder<T> coder;
+    private String coderName;
+
+    public CoderTypeSerializerConfigSnapshot() {
```
Yeah, here to solve BEAM-2407.

JingsongLi commented on runners/flink/src/main/java/org/apache/beam/runners/flink/translation/types/CoderTypeSerializer.java:
```diff
@@ -139,24 +140,28 @@ public TypeSerializerConfigSnapshot snapshotConfiguration() {
-   *  TypeSerializerConfigSnapshot of CoderTypeSerializer.
+   *  TypeSerializerConfigSnapshot of CoderTypeSerializer. This uses the class name of the
+   *  {@link Coder} to determine compatibility. This is a bit crude but better than using
+   *  Java Serialization to (de)serialize the {@link Coder}.
```
Why better than using Java Serialization? Performance reasons?

aljoscha replied:
In Flink, there has been a trend to move as far away from Java Serialization as possible because it can lead to problems in deserialisation when the code is changed.

Simply relying on the class name is also not the best solution, I admit. At some point there might be a Coder versioning API in Beam. @kennknowles Do you have any comments? I think you once mentioned that you had some work on coder versioning.

JingsongLi commented on runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/io/UnboundedSourceWrapper.java:
```diff
@@ -283,6 +283,8 @@ public void run(SourceContext<WindowedValue<ValueWithRecordId<OutputT>>> ctx) th
         }
       }
 
+      setNextWatermarkTimer(this.runtimeContext);
```
Yeah, here to solve BEAM-2408.

JingsongLi commented on runners/flink/src/test/java/org/apache/beam/runners/flink/streaming/UnboundedSourceWrapperTest.java:
```diff
@@ -173,6 +168,106 @@ public void close() {
+
+                  @Override
+                  public void emitWatermark(Watermark watermark) {
+                    if (watermark.getTimestamp() >= numElements / 2) {
```
Why is numElements / 2 instead of numElements?

aljoscha replied:
I just picked some arbitrary watermark between `-Inf` and numElements, numElements works as well.

JingsongLi replied:
Yes, I know what you mean. Just seenWatermark.

JingsongLi commented on runners/flink/src/test/java/org/apache/beam/runners/flink/streaming/UnboundedSourceWrapperTest.java:
```diff
@@ -173,6 +168,106 @@ public void close() {
+      // with the outer Thread
+      final AtomicBoolean seenWatermark = new AtomicBoolean(false);
+
+      new Thread() {
```
Should we shutdown this thread?

aljoscha replied:
I was lazy but you're absolutely right. I'll store the Thread in a variable and add this at the end:
```
sourceOperator.cancel();
sourceThread.join();
```

aljoscha commented:
Run Flink ValidatesRunner

aljoscha commented:
@JingsongLi what do you think about merging this?

JingsongLi commented:
LGTM. 👍
